### ISSUE TITLE
Always throwing Blessed.LaunchdError error 3 - authorizationFailed

### ISSUE THREAD

**Issue by ha1lie** [trilemma-dev/Blessed#12]
Hello!

I'm attempting to use this as a part of an application running a privileged helper, and am using [SwiftAuthorizationSample](https://github.com/trilemma-dev/SwiftAuthorizationSample) as a jumping off point. I've been having issues getting this to function in my own app, so here is a short summary:

I started off by running the SwiftAuthorizationSample project, just changing the Signing ID to my developer ID. When running the project, it worked flawlessly, installed the helper tool, and I could see it located in /Library/PrivilegedHelperTools. Success!

So, I started initially by going through the code in that project to create my own version which accomplishes running a privileged command in a different communication method. This is when I first ran into the problem of it not installing. It threw a `Blessed.LaunchdError error 3`. This would come to haunt me.

My initial thought was that something in my implementation had been wrong. So, out of curiousity, I began by recreating the helper tool, and the shared resources in my project, knowing that if I can get those to work, it's my implementation, and not my project files. I dragged and dropped, verified that my `Info.plist`s were correct, run scripts, even scrolling through Build Settings side by side to make sure they matched. 

*No such luck.* Ouch.

So then I tried to back trace where this error could be from. When I followed where it was running, I figured that it had to be an error thrown at line 81 in `LaunchdManager.swift` in the Blessed package. I couldn't find anywhere explicitly throwing an authorizationError, soooooo.

I then figured I'd take a look at the errors in both this project, as well as [SwiftAuthorizationSample](https://github.com/trilemma-dev/SwiftAuthorizationSample), and the closest issues I could find are [#8](https://github.com/trilemma-dev/SwiftAuthorizationSample/issues/8) and [#6](https://github.com/trilemma-dev/SwiftAuthorizationSample/issues/6) both of which seem like they should have been resolved by [this solution](https://github.com/trilemma-dev/SwiftAuthorizationSample/commit/1446ed3817b4ce25497b7ae381374b24d2e4cdd0). The root of the problem seemed to be addressed, which was that the Signing ID's needed to be in quotes. As my code pull was after this change, I'm not sure what could be the issue. 

When I ran the install code, it showed a dialogue, I entered my password, and without a doubt, every time, it would throw that error. I double verified everything again. After all, it has to be me. The only issue that it could be is that the app may not have permissions to write to `/Library`, but the `.xcconfig` states that it should not be sandboxed, and I was unable to get the NSApplication.isSandboxed() check to return anything, as it always just threw an error when running. 

I've done my best to ensure that all of the settings are true to what they should be, the project is indeed using the correct `.xcconfig` files, and all information is true. It does in fact copy the helper tool to `DerivedData/Build/Products/Debug/Maple/Contents/Library/LaunchServices/dev.halz.Maple.helper`. This is the output of `DiagnosticSigningInfo.printDiagnosticInfo()`:

```
==============This App=============
2022-05-10 09:41:30.260574-0400 Maple[5475:141871] [logging-persist] cannot open file at line 45530 of [9ff244ce07]
2022-05-10 09:41:30.260631-0400 Maple[5475:141871] [logging-persist] os_unix.c:45530: (0) open(/var/db/DetachedSignatures) - Undefined error: 0
CFBundleIdentifier: dev.halz.Maple
CFBundleVersion: 1
Common Name: Apple Development: My Name (CT44P6LGD8)
Organizational Unit: 5RD23JCCKU

========Bundled Helper Tool========
CFBundleIdentifier: dev.halz.Maple.helper
CFBundleVersion: 1.0.1
Common Name: Apple Development: My Name (CT44P6LGD8)
Organizational Unit: 5RD23JCCKU

=======Installed Helper Tool=======
Helper tool is not installed
```

I appreciate any input you have in this issue! It's much appreciated.

**Comment by jeff-h** [comment]
Sad to hear of your troubles. My experience with privileged helpers is a) they can be a colossal pain to get working and b) the trilemma-dev libraries help ease that pain enormously. So I guess you can be assured you're in the right place at least, and that you're not alone in your journey :)

I'm guessing you have already checked each of the eleven items in the checklist at the top of LaunchdManager. Still might be worth doing with fresh eyes.

3: 
`codesign --verify -V <path to app bundle>`
`codesign --verify -V <path to app bundle's helper binary>`

6–10: I don't believe you can view the embedded launchd info plist of a commandline tool, but this will show you the embedded info.plist:
`launchctl plist <path to app bundle's helper binary>`

Also, might be worth checking your helper is "alive" — if you added some command line code (eg [this from the example code](https://github.com/trilemma-dev/SwiftAuthorizationSample/blob/2881559954d8c5f90d772e1d13c0e5ba8627ce96/SwiftAuthorizationHelperTool/main.swift#L15-L27)), you can run your helper on the command line:

```
$ cd <into your helper location inside your app bundle>
$ ./dev.halz.Maple.helper
```

Is your app just a dummy test at this stage? If so, perhaps put it on GitHub somewhere — I'd be happy to see if anything jumps out to me.

**Comment by ha1lie** [comment]
Thanks for your input! I've found this has so far been the most difficult part of my project! 

I have done my best on my own to verify that the eleven items mentioned in the original project's README. (I only put this issue on the Blessed framework as I figured it applies more to the actual *blessing* and not the helper tool). However, going through the pointers that you included:

3: Using codesign commands to verify states that both the app, and it's contained helper tool, are valid, and satisfy all requirements. 

6-10: As far as I can tell(by removing the scripts to clean up the `info.plist`s), it is at least doing those. Obviously I can't tell the launchd.plist is being generated correctly, but I must only assume so. Running the command you gave me to verify the info.plist of the helper tool, I can verify that it is present, and accurate, including matching signing ID. 

I'm not so sure what you mean as far as being "alive", however, I can run the executable from the command line, and it spits back out the usage command to me(the current helper implementation is just verbatim what is in the SwiftAuthorizationSample). So, it functions. At least to the basics that it is needed to. 

After looking into it a bit more, my thoughts are one of the following things:
1. I cannot write the file to `/Library/PrivilegedHelperTools` - I attempted to write just a text file to the same folder, but I cannot. However, part of me feels like the need to be `root` to write to the folder should be solved by `SMJobBless`
2. (And this one is a sure thing) SMJobBless is failing. I figured this out by including the Blessed framework as a local code package instead of an SPM package, and adding logs in the functions being called by `authorizeAndBless`. Without fail, every time, it throws: `CFErrorDomainLaunchd error 4.`, which is obviously not an accessuble thing to figure out where the deeper error is. 

The project I'm trying to get this to work for wasn't public, but will eventually be, and there isn't too much in there that is "sensitive", and it can't be truly utilized without also having a developer framework, so it is now a **public** repository, as I need help with this, and I'm not quite sure what is causing the issue. You can see it [here](https://github.com/ha1lie/Maple)

I appreciate your help and input!

**Comment by jakaplan** [comment]
Sorry to hear you're struggling with blessing a helper tool, it is unfortunately quite a pain to get working despite my best efforts so far. `Blessed` is essentially just wrapper around Apple's frameworks. Unfortunately the `LaunchdError.authorizationFailure` error which corresponds to [`kSMErrorAuthorizationFailure`](https://developer.apple.com/documentation/servicemanagement/ksmerrorauthorizationfailure)'s documentation is bad or at least woefully understated. Despite what the documentation claims, in practice it appears to mean "I failed for some reason which I shall not explain to you".

Looking at the diagnostic info you provided, it looks like you've misconfigured usage of one or more of the `xcconfig` files. The [relevant portion](https://github.com/trilemma-dev/SwiftAuthorizationSample/blob/2881559954d8c5f90d772e1d13c0e5ba8627ce96/Config.xcconfig#L16-L18):
```
// The app's version needs to be defined at the project level so the helper tool can encode this as a code requirement
// mitigate downgrade attacks. See README for details.
APP_VERSION = 1.0.0
```

Whereas your diagnostic output is:
```
==============This App=============
CFBundleIdentifier: dev.halz.Maple
CFBundleVersion: 1
```

This does not match, it should say `CFBundleVersion: 1.0.0`.

When `PropertyListModifier.swift` generates the requirements for what application can install it, it's using the build variable set by the `Config.xcconfig` file. So it's setting the requirement that an application with `CFBundleVersion` of `1.0.0` or greater can install (and update) it. I'm pretty sure an application with a `CFBundleVersion` of `1` is not equal to or greater, as unintuitive as that might seem.

Even if that's not the exact issue, there's probably another issue or two coming from mismatches as a result of this misconfiguration.

**Comment by ha1lie** [comment]
That was.... that was it. I would honestly have driven myself crazy trying to find that, and never would have even thought it would trip up on that. I am very very appreciative to both you and jeff-h for your help! My apologies for opening up an issue which maybe shouldn't have been! It installs fine now, and I'm able to communicate with it as I expected. I'll close this issue now :)

**Comment by jakaplan** [comment]
Glad to be of help. I have aspirations to eventually be able to detect issues like this and have the Blessed framework generate much more specific errors with helpful guidance. However, to do so comprehensively I think I'd end up needing to write my own parser for the Code Signing Requirement language. Which is...involved to say the least.

**Comment by jakaplan** [comment]
> I was unable to get the NSApplication.isSandboxed() check to return anything, as it always just threw an error when running.

This has been fixed in the just released [0.4.0](https://github.com/trilemma-dev/Blessed/releases/tag/0.4.0) version.
